packages/salesforcedx-vscode-replay-debugger/src/states/variableAssignmentState.ts

```diff
@@ -26,106 +26,161 @@ export class VariableAssignmentState implements DebugLogState {
         }
-        if (value.indexOf('{') === 0 && value !== '{}') {
-          container.value = '';
-          container.variablesRef = logContext
-            .getVariableHandler()
-            .create(container);
-          this.parseAndPopulate(value, container);
+        const refContainer = refMap.get(ref)!;
+        // nested variable will either be given a json or a value
+        if (isNested) {
+          if (value === '{}') {
+            refContainer.variables.set(
+              varName,
+              new ApexVariableContainer(varName, value, '')
+            );
+            // if its a json assignment, parse the values
+          } else if (value.indexOf('{') === 0) {
+            const topLevel = new ApexVariableContainer(varName, '', '');
+            refContainer.variables.set(varName, topLevel);
+            topLevel.variablesRef = logContext
+              .getVariableHandler()
+              .create(topLevel);
+            this.parseJSONAndPopulate(value, topLevel, logContext);
+          } else {
+            // if it's not nested then we check if the value is a reference
+            if (refMap.has(value)) {
+              const pulledRef = refMap.get(value) as ApexVariableContainer;
+              const tmpContainer = new ApexVariableContainer(
+                varName,
+                pulledRef.value,
+                pulledRef.type,
+                value
+              );
+              tmpContainer.variables = pulledRef.variables;
+              tmpContainer.variablesRef = logContext
+                .getVariableHandler()
+                .create(tmpContainer);
+              refContainer.variables.set(varName, tmpContainer);
+              // if not a reference, update the variable value, creating a container if needed
+            } else if (refContainer.variables.has(varName)) {
+              const varContainer = refContainer.variables.get(
+                varName
+              ) as ApexVariableContainer;
+              varContainer.value = value;
+            } else {
+              refContainer.variables.set(
+                varName,
+                new ApexVariableContainer(varName, value, '')
+              );
+            }
+          }
+          // if not nested then the refcontainer is the top level
+        } else if (value.indexOf('{') === 0) {
+          this.parseJSONAndPopulate(value, refContainer, logContext);
+        } else {
+          refContainer.variables.set(
+            varName,
+            new ApexVariableContainer(varName, value, '')
+          );
         }
-        // assigning to variable's fields, currently only going to work for a variable in local
-      } else if (name.indexOf('.') !== -1 && ref !== '0') {
-        container = logContext.getRefsMap().get(ref)!;
-        if (container) {
+
+        // update toplevel container if it's not this and not a collection
+        // or if the this variable has not been assigned a reference yet
+        if (
```

Comment:
@DatGuyJonathan I just now added this in because I realized that the current frame's local `this` would be overwritten sometimes if nested variables were changed in a different object. This will just check that if we are updating `this`, the container is only updated if this is the first assignment.